dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/registry/impl/ServiceRegistryActor.scala

```diff
@@ -19,100 +19,182 @@ import scala.collection.JavaConverters._
-          else
-            sender() ! Status.Failure(new ServiceAlreadyRegistered(name))
+  def register(serviceName: String, details: ServiceRegistryService): Any = {
+    val subset: Map[ServiceRegistryKey, (URI, ServiceRegistryService)] = reg.filter { case (k, _) => k.serviceName.name == serviceName }
+    if (subset.isEmpty) {
+      if (logger.isDebugEnabled) {
+        logger.debug(s"Registering service [$serviceName] with ACLs [${details.acls().asScala.map { acl => acl.toString }.mkString(", ")}] on ${details.uris().asScala.mkString(",")}).")
       }
-    case GetRegisteredServices =>
-      val services: List[RegisteredService] = (for {
-        (name, service) <- registry
-      } yield RegisteredService.of(name, service.uri))(collection.breakOut)
-      import scala.collection.JavaConverters._
-      sender() ! RegisteredServices(TreePVector.from(services.asJava))
-    case route: Route =>
-      sender() ! router.lift(route).fold[RouteResult](NotFound(registry))(Found.apply)
-      warnOnAmbiguity(route)
+      reg = reg ++ InternalRegistry.build(serviceName, details)
+      Done
+    } else {
+      val actualDetails: ServiceRegistryService = subset.values.map(_._2).head
+      if (actualDetails.equals(details)) {
+        Done // idempotent, same already registered
+      } else {
+        Status.Failure(new ServiceAlreadyRegistered(serviceName))
+      }
+    }
   }
 
-  private def warnOnAmbiguity(route: Route) = {
-    if (logger.isWarnEnabled) {
-      val servingRoutes = routerFunctions.filter(_.isDefinedAt(route))
-      if (servingRoutes.size > 1) {
-        val servers = servingRoutes.map(_.apply(route))
-        logger.warn(s"Ambiguous route resolution serving route: $route. Route served by ${servers.head} but also matches ${servers.tail.mkString("[", ",", "]")}.")
-      }
+  def remove(serviName: String) =
+    reg = reg.filterNot { case (k, _) => k.serviceName.name == serviName }
+}
+
+class InternalRouter {
+  private val logger: Logger = Logger(classOf[InternalRouter])
+  import ServiceRegistryActor._
+
+  private var router = PartialFunction.empty[Route, (URI, ServiceRegistryService)]
+  // maps a Route to ServiceNames
+  private var routerFunctions = Seq.empty[PartialFunction[Route, (URI, ServiceRegistryService)]]
+  private var simpleRegistry: Map[String, (URI, ServiceRegistryService)] = Map.empty[String, (URI, ServiceRegistryService)]
+
+  def routeFor(route: ServiceRegistryActor.Route): RouteResult = {
+    router.lift(route).fold[RouteResult](NotFound(simpleRegistry)) {
+      case (uri, _) =>
+        Found(uri)
     }
   }
 
-  private def serviceRouter(service: ServiceRegistryService) = {
-    val addressUri = service.uri
+  def rebuild(registry: InternalRegistry): Unit = {
+    routerFunctions = registry.reg.map { case (k, v) => (k.serviceName.name, k.portName, v._1, v._2) }.toSeq.flatMap { case (name, portname, uri, details) => serviceRouter(name, portname, uri, details) }
```

Comment:
This could be breakdown in a few lines and with some comments. 
Hard to read such a long line.